[CI] SyntheticVersusColumnarStoredSourceIT testStoredIgnoredSource failing

**Build Scans:**
- [elasticsearch-intake #42954 / part1](https://gradle-enterprise.elastic.co/s/o2zc6wth3vqfy)

**Reproduction Line:**
```
./gradlew ":server:internalClusterTest" --tests "org.elasticsearch.index.mapper.SyntheticVersusColumnarStoredSourceIT.testStoredIgnoredSource" -Dtests.seed=FB130889FCD24419 -Dtests.locale=ii-CN -Dtests.timezone=Africa/Maputo -Druntime.java=26
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.index.mapper.SyntheticVersusColumnarStoredSourceIT),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('testStoredIgnoredSource'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: expected:<{HQBK=-1334998488865281176, Vw=[{QBOWQ=[XoXknglwdAuEaVIleSpkliDOgpKooLKwVUbOfKNGeOw, KnmZGfgvmkReYeZjBoAoSl, SXfhmmKRAASGjsUcrgsXKYEeghnNIffQfgCBIVRKOMYutCh], XjT=0.42403919, eQSKahMWC=kZnqe}, {QBOWQ=esDzUFrnjmJJYWLDuWXzuGli, jGp=0.35376897}, {QBOWQ=eNMjbmhaLGbKe, XjT=[0.7843479, 0.23123854], eQSKahMWC=[iQjoDu, WxbKAZzhUyIchImzUGu, OuOttFdLrvZAU, wdlVEruuSsbmrUGpBWGUXYlyRgHINxngkKdc, ZOxDIGpFHkQVvVdTcyyFXiozGtXIedvwCdxLLGw], jGp=0.48620218}, {QBOWQ=wEaoAoRNCmnTeTEUkLIWNuNfUwwzOorxyOtcQqKVGFadyLXWKf, XjT=0.50222206, eQSKahMWC=[xbTTDqmcYxeqvTAQbcHniXsHDDi, twLEgxbUfbPYGPZJlWsYNfCwXxUSSsrJtlxVZozDj, AKLjL], jGp=[0.9664106, 0.8269716, 0.058812574, 0.177188, 0.24343595]}]}> but was:<{HQBK=-1334998488865281176, Vw=[{QBOWQ=[XoXknglwdAuEaVIleSpkliDOgpKooLKwVUbOfKNGeOw, KnmZGfgvmkReYeZjBoAoSl, SXfhmmKRAASGjsUcrgsXKYEeghnNIffQfgCBIVRKOMYutCh], XjT=0.42403919, eQSKahMWC=kZnqe}, {QBOWQ=[XoXknglwdAuEaVIleSpkliDOgpKooLKwVUbOfKNGeOw, KnmZGfgvmkReYeZjBoAoSl, SXfhmmK
[truncated]
```

**Issue Reasons:**
- [main] 1 consecutive failures in test testStoredIgnoredSource
- [main] 1 consecutive failures in step part1
- [main] 1 consecutive failures in pipeline elasticsearch-intake

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -764,9 +764,6 @@ tests:
 - class: org.elasticsearch.xpack.esql.CsvIT
   method: test {csv-spec:exponential_histogram.allAggsInlineGrouped}
   issue: https://github.com/elastic/elasticsearch/issues/153123
-- class: org.elasticsearch.index.mapper.SyntheticVersusColumnarStoredSourceIT
-  method: testStoredIgnoredSource
-  issue: https://github.com/elastic/elasticsearch/issues/153124
 - class: org.elasticsearch.xpack.stateless.recovery.PointInTimeRelocationIT
   method: testPointInTimeRelocationPitPositionsInBcc
   issue: https://github.com/elastic/elasticsearch/issues/153149

--- a/server/src/internalClusterTest/java/org/elasticsearch/index/mapper/SyntheticVersusColumnarStoredSourceIT.java
+++ b/server/src/internalClusterTest/java/org/elasticsearch/index/mapper/SyntheticVersusColumnarStoredSourceIT.java
@@ -132,6 +132,35 @@ public class SyntheticVersusColumnarStoredSourceIT extends ESIntegTestCase {
         assertEqualSource(mappingXContent, document, randomBoolean());
     }
 
+    /**
+     * With the time-series doc-values format disabled, a keyword sub-field of a nested field whose values exceed
+     * {@code ignore_above} keeps a copy of the ignored values in a per-document stored field so synthetic source can reconstruct
+     * them. Each nested array entry is a separate Lucene document, but columnar_stored reconstructs them all through one reused
+     * single-document reader that always reports the same doc id, and a reused stored-field loader skips re-reading on an unchanged
+     * doc id - so it used to return the first entry's stored values for every sibling. Verify every entry keeps its own values.
+     */
+    public void testNestedWithIgnoredKeywordPerEntry() throws Exception {
+        var mappingXContent = XContentFactory.jsonBuilder()
+            .startObject()
+            .startObject("properties")
+            .startObject("n")
+            .field("type", "nested")
+            .startObject("properties")
+            .startObject("kw")
+            .field("type", "keyword")
+            .field("ignore_above", 4)
+            .endObject()
+            .endObject()
+            .endObject()
+            .endObject()
+            .endObject();
+        // Every value exceeds ignore_above (length 4), so all are stored in the ignored-value fallback. The first entry is
+        // multi-valued to make first-entry-duplication observable if the reused stored-field loader is not read per entry.
+        var document = Map.of("n", List.of(Map.of("kw", List.of("aaaaa", "bbbbb")), Map.of("kw", "ccccc"), Map.of("kw", "ddddd")));
+        // The stored-field fallback (and thus this bug) only occurs with the time-series doc-values format disabled.
+        assertEqualSource(mappingXContent, document, false);
+    }
+
     private void runTest(boolean useTimeSeriesDocValuesFormat) throws Exception {
         var spec = buildSpec();
         var template = new TemplateGenerator(spec).generate();

--- a/server/src/main/java/org/elasticsearch/index/mapper/NestedObjectMapper.java
+++ b/server/src/main/java/org/elasticsearch/index/mapper/NestedObjectMapper.java
@@ -476,7 +476,6 @@ public class NestedObjectMapper extends ObjectMapper {
         private boolean columnar;
         private ColumnarSourceWriter.ReusableColumnarStoredLeafReader columnarChildReader;
         private SourceLoader.Leaf columnarChildLeaf;
-        private LeafStoredFieldLoader columnarChildStoredLoader;
         private final List<LuceneDocument> columnarChildren = new ArrayList<>();
 
         private NestedSyntheticFieldLoader(
@@ -502,10 +501,6 @@ public class NestedObjectMapper extends ObjectMapper {
                 this.columnarChildren.clear();
                 this.columnarChildReader = new ColumnarSourceWriter.ReusableColumnarStoredLeafReader();
                 this.columnarChildLeaf = sourceLoader.leaf(columnarChildReader, ColumnarSourceWriter.DOC_IDS);
-                this.columnarChildStoredLoader = storedFieldLoader.getLoader(
-                    columnarChildReader.getContext(),
-                    ColumnarSourceWriter.DOC_IDS
-                );
                 return parentDoc -> {
                     columnarChildren.clear();
                     LuceneDocument parent = parentReader.currentDoc();
@@ -590,8 +585,10 @@ public class NestedObjectMapper extends ObjectMapper {
 
         private void writeColumnarChild(LuceneDocument child, XContentBuilder b) throws IOException {
             columnarChildReader.repopulate(child);
-            columnarChildStoredLoader.advanceTo(ColumnarSourceWriter.DOC_ID);
-            columnarChildLeaf.write(columnarChildStoredLoader, ColumnarSourceWriter.DOC_ID, b);
+            // A fresh loader per child: every child reuses DOC_ID (0), and a reused loader would skip re-reading stored fields.
+            var childStoredLoader = storedFieldLoader.getLoader(columnarChildReader.getContext(), ColumnarSourceWriter.DOC_IDS);
+            childStoredLoader.advanceTo(ColumnarSourceWriter.DOC_ID);
+            columnarChildLeaf.write(childStoredLoader, ColumnarSourceWriter.DOC_ID, b);
         }
 
         /**

